### PR TITLE
cpu/stm32: add pm support in uart, spi and i2c (f4)

### PR DESCRIPTION
**Pull request by vincent-d**
This add `pm_layered`-based power management in uart, spi and i2c (f4 only so far though).

### CODE REVIEW

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 1]
```diff
@@ -75,6 +75,14 @@ extern "C" {
 #if defined(CPU_FAM_STM32F1) || defined(CPU_FAM_STM32F2) \
     || defined(CPU_FAM_STM32F4) || defined(DOXYGEN)
 #define PM_NUM_MODES    (2U)
+
+/**
+ * @name    Power modes
+ * @{
+ */
+#define PM_STOP         (1U)
```
Comment (roberthartung) [inline cpu/stm32_common/include/periph_cpu_common.h]: `/cpu/stm32_common/periph/pm.c` should use these defines in `pm_set`.

Comment (vincent-d) [reply cpu/stm32_common/include/periph_cpu_common.h]: Fixed

**Comment by vincent-d** [comment]
Changed uart pm so that stop mode is blocked only if rx is used

**Comment by olegart** [comment]
I don't quite understand the concept.

E.g. we have sys/shell enabled — so it will block all power management until we manually disable corresponding UART every time before we go to sleep and reenable it right after wakeup?

Seems totally meaningless to me. Why anyone needs it? If you need shell access all the time, just disable PM at all.

**Comment by roberthartung** [comment]
@olegart a program with shell will never sleep, that's correct.

**Comment by kaspar030** [comment]
> I don't quite understand the concept.

I think you do. The concept is just not fully implemented, thus you're right, on most (if not all) platforms, simple UART RX will block most low power modes.

**Comment by olegart** [comment]
Yes, but why?

If I need permanent shell access, I'd rather disable LPM at all, no difference with current RIOT logic. If I need sleepy device with occasional shell access (e.g. we have devices with 15 seconds of regular shell right after boot), I'll disable/enable LPM when I need that access.

With proposed changes — nothing actually changes for us in the first case, and the second case just doesn't work anymore as I can't just disable UART peripheral current in use by the shell.

**Comment by olegart** [comment]
Long story short,

1) without layered PM — I can use shell (UART) whenever I want to by disabling LPM and I can have some sleep whenever I want to by enabling LPM

2) with layered PM — I can't use shell and LPM on the same system, period.

That's not an added functionality, that's reduced functionality.

**Comment by kaspar030** [comment]
> we have devices with 15 seconds of regular shell right after boot

You can probably still do that. How's the timeout currently implemented?

**Comment by kaspar030** [comment]
> with layered PM — I can't use shell and LPM on the same system, period.

That is not correct.

**Comment by olegart** [comment]
> How's the timeout currently implemented?

RTC timer setting pm_prevent_sleep global variable to 0 after 15 seconds. PM checks the variable before switching to STOP mode.

> You can probably still do that.

No. There's no option to disable shell, shell_run function is a one-way ticket. If I power down UART port used by it, shell will hang on the very next attempt to TX anything.

**Comment by kaspar030** [comment]
> There's no option to disable shell, shell_run function is a one-way ticket. If I power down UART port used by it, shell will hang on the very next attempt to TX anything.

Let's please not mix up concepts and implementation.

Your use case could probably be implemented in uart_stdio (e.g., by adding ```uart_stdio_enable/disable_rx()```), which would reinitialize the underlying uart with/without rx callback. Or in periph/uart, by offering a function to set/clear the rx uart independently from complete (re-)initialization.

You could also just unblock the corresponding powermode on your board once, after your timeout. Or add an option to shell to exit after a certain time.

The implementation of pm_layered is currently incomplete, but the general concept seems to map to most use-cases, on most hardware, so far.

**Comment by roberthartung** [comment]
See also: #7947

**Comment by kaspar030** [comment]
Please squash!

**Comment by vincent-d** [comment]
rebased and squashed

**Comment by kaspar030** [comment]
Due tue the default pm_blocker, problems won't show up for a while... ;)

**Pull request merged**